src/kernel/scheduling/task.c

```diff
@@ -1,46 +1,97 @@
+            size_t vmm_hdr_flags = elf_flags_to_memory_flags(phdr.p_type);
+            pretty_logf(Verbose, "\t[%d]: Type: 0x%x, Flags: 0x%x  -  Vaddr: 0x%x - aligned: 0x%x ", i, phdr.p_type, phdr.p_flags, phdr.p_vaddr, align_value_to_page(phdr.p_vaddr));
+            pretty_logf(Verbose, "\t\t - FileSz: 0x%x, Memsz: 0x%x, vmm flags: 0x%x - p_align: 0x%x - p_offset: 0x%x", phdr.p_filesz, phdr.p_memsz, vmm_hdr_flags, phdr.p_align, phdr.p_offset);
+            Elf64_Half mem_pages = (Elf64_Half) get_number_of_pages_from_size(phdr.p_memsz);
+            Elf64_Half filesz_pages = (Elf64_Half) get_number_of_pages_from_size(phdr.p_filesz);
+            uint64_t offset_address = (uint64_t) ((uint64_t) elf_header + (uint64_t) phdr.p_offset);
+            uint64_t vaddr_address = align_value_to_page(phdr.p_vaddr);
+            for (int j = 0; j < mem_pages; j++) {
+                pretty_logf(Verbose, "[%d]: Mapping: offset: 0x%x (virtual: 0x%x) in vaddr: 0x%x", j, hhdm_get_phys_address((uintptr_t) offset_address), offset_address, vaddr_address);
+                map_phys_to_virt_addr_hh(hhdm_get_phys_address(offset_address), (void *) vaddr_address, VMM_FLAGS_USER_LEVEL | vmm_hdr_flags | VMM_FLAGS_PRESENT, (uint64_t *) new_task->vmm_data.root_table_hhdm);
+                //I need a mem copy. I need to fopy the content of elf_header + phdr.p_offset into phdr.p_vaddr
+                offset_address += (uint64_t) phdr.p_align;
+                vaddr_address += (uint64_t) phdr.p_align;
+                //I need to allocate memory and map it into the new memory space,
+            }
+        }
+        thread_t* thread = create_thread(name, (void (*)(void *))elf_header->e_entry, args, new_task, false, true);
+        pretty_logf(Verbose, "Thread created: id: %d - Entry Point: 0x%x - elf header entry point: 0x%x", thread->tid, thread->execution_frame->rip, elf_header->e_entry);
+    }
+    // Create a new thread
+    scheduler_add_task(new_task);
+    asm("sti");
+    return new_task;
+}
+
+task_t *create_task_from_func(char *name, void (*_entry_point)(void *), void *args, bool is_supervisor) {
+    asm("cli");
+    task_t* new_task = create_task(name,  is_supervisor);
+    prepare_virtual_memory_environment(new_task);
+        if( is_supervisor) {
         pretty_logf(Verbose, "creating new supervisor thread: %s", name);
-        thread_t* thread = create_thread(name, _entry_point, args, new_task, is_supervisor);
+        thread_t* thread = create_thread(name, _entry_point, args, new_task, is_supervisor, false);
         new_task->threads = thread;
     } else {
         pretty_logf(Verbose, "creating new userspace thread %s", name);
-        thread_t* thread = create_thread(name, NULL, args, new_task, is_supervisor);
+        thread_t* thread = create_thread(name, NULL, args, new_task, is_supervisor, false);
         new_task->threads = thread;
     }
     scheduler_add_task(new_task);
-    //re-enable interrupts
     asm("sti");
```

Comment:
you enable interrupts unconditionally here, what if they were previously disabled before calling this function? Now the caller thinks they are disabled but in reality they arent. This is a pattern that will be used a lot, so it can be helpful to have something like:
```c
const bool restore_intrs = disable_interrupts(); //returns previous state and ensure interrupts are disabled
[ ... ]
if (restore_intrs)
    enable_interrupts();
```